File: Homework/quoc_khanh/lesson_11/18.4Sum.py
from typing import List

class Solution:
    def twoSum(self, nums:List[int], target: int) -> List[List[int]]:
        result = []
        left, right = 0, len(nums) - 1

        def decreaseRight():
            nonlocal left, right
            while left < right and nums[right] == nums[right - 1]:
                right -= 1
            right -= 1
        
        def increaseLeft():
            nonlocal left, right
            while right > left and nums[left + 1] == nums[left]:
                left += 1
            left += 1

        while left < right:
            leftNum, rightNum = nums[left], nums[right]
            currSum = leftNum + rightNum
            if currSum == target:
                result.append([leftNum, rightNum])
                increaseLeft()
                decreaseRight()
            if currSum > target:
                decreaseRight()
            if currSum < target:
                increaseLeft()
        return result

    def kSum(self, nums: List[int], target: int, k: int) -> List[List[int]]:
        if k == 2:
            return self.twoSum(nums, target)
        length = len(nums)
        result = []
        i = 0
        while i <= length - k:
            currentNum = nums[i]
            results = self.kSum(nums[i+1:], target - currentNum, k - 1)
            for res in results:
                result.append([currentNum, *res])
            while i < length - k and nums[i] == nums[i+1]:
                i += 1
            i += 1
        return result

    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        return self.kSum(nums, target, 4)

s = Solution()
# print(s.fourSum([1,0,-1,0,-2,2], 0) == [[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]])
print(s.fourSum([2,2,2,2], 8) == [[2,2,2,2]])

print(s.twoSum([1,2,3,4,5], 6) == [[1, 5],[2,4]])
print(s.twoSum([1,1,2,2,3,3,4,4,5,5], 6) == [[1, 5], [2, 4], [3, 3]])
print(s.twoSum([2,2,2,2,2,2], 4) == [[2,2]])
print(s.twoSum([2,2], 4) == [[2,2]])

File: Homework/Hoang_Nguyen/lesson11/18_4Sum.js
var twoSum = function (nums, left, target) {
  const res = [];
  let right = nums.length - 1;
  while (left < right) {
    const sum = nums[left] + nums[right];
    if (sum < target) {
      left++;
    } else if (sum > target) {
      right--;
    } else {
      res.push([nums[left], nums[right]]);
      // prevent duplication
      while (left < nums.length - 1 && nums[left] === nums[left + 1]) {
        left++;
      }
      while (right > 0 && nums[right] === nums[right - 1]) {
        right--;
      }
      left++;
      right--;
    }
  }
  return res;
};

var threeSum = function (nums, left, target) {
  const res = [];
  for (let i = left; i <= nums.length - 2; i++) {
    // prevent duplication
    if (i > left && nums[i] === nums[i - 1]) {
      continue;
    }
    const twoSumRes = twoSum(nums, i + 1, target - nums[i]);
    twoSumRes.forEach((arr) => res.push([nums[i], ...arr]));
  }
  return res;
};

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function (nums, target) {
  nums.sort((a, b) => a - b);
  const res = [];
  for (let i = 0; i < nums.length - 3; i++) {
    // prevent duplication
    if (i > 0 && nums[i] === nums[i - 1]) {
      continue;
    }
    const threeSumRes = threeSum(nums, i + 1, target - nums[i]);
    threeSumRes.forEach((arr) => res.push([nums[i], ...arr]));
  }
  return res;
};
